pkg/machine/machineset.go

```diff
@@ -0,0 +1,378 @@
+			"Error initializing MachineSet from copy: %s", err.Error())
+
+		builder.errorMsg = fmt.Sprintf("Error initializing MachineSet from copy: %s",
+			err.Error())
+	}
+
+	if nsName == "" {
+		glog.V(100).Infof("The Namespace of the MachineSet is empty")
+
+		builder.errorMsg = "MachineSet 'nseName' cannot be empty"
+	}
+
+	if instanceType == "" {
+		glog.V(100).Infof("The instanceType of the MachineSet is empty")
+
+		builder.errorMsg = "MachineSet 'instanceType' cannot be empty"
+	}
+
+	if replicas == 0 {
+		glog.V(100).Infof("The replicas of the MachineSet is zero")
+
+		builder.errorMsg = "MachineSet 'replicas' cannot be zero"
+	}
+
+	if builder.Definition == nil {
+		glog.V(100).Infof("The MachineSet object definition is nil")
+
+		builder.errorMsg = "MachineSet 'Object.Definition' is nil"
+	}
+
+	return &builder
+}
+
+// PullMachineSet loads an existing MachineSet into Builder struct.
+func PullMachineSet(apiClient *clients.Settings, name, namespace string) (*MachineSetBuilder, error) {
+	glog.V(100).Infof("Pulling existing machineSet name %s in namespace %s", name, namespace)
+
+	builder := MachineSetBuilder{
+		apiClient: apiClient,
+		Definition: &machinev1beta1.MachineSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+		},
+	}
+
+	if name == "" {
+		builder.errorMsg = "MachineSet 'name' cannot be empty"
+	}
+
+	if namespace == "" {
+		builder.errorMsg = "MachineSet 'namespace' cannot be empty"
+	}
+
+	if !builder.Exists() {
+		return nil, fmt.Errorf("MachineSet object %s doesn't exist in namespace %s", name, namespace)
+	}
+
+	builder.Definition = builder.Object
+
+	return &builder, nil
+}
+
+// CreateNewWorkerMachineSetFromCopy returns a MachineSet object which was copied from an exiting one on cluster.
+func CreateNewWorkerMachineSetFromCopy(
+	apiClient *clients.Settings,
+	namespace,
+	publicCloud,
+	instanceType string,
+	replicas int32) (*machinev1beta1.MachineSet, error) {
+	// currently only supporting AWS MachineSets
+	if publicCloud != "aws" {
+		return nil, fmt.Errorf("only AWS public cloud is currently supported for copied MachineSet")
+	}
+
+	workerMachineSetList, err := GetWorkerMachineSetList(apiClient, namespace)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not get worker MachineSetList: %w", err)
+	}
+
+	// picking the first worker MachineSet in list
+	baseMs := workerMachineSetList.Items[0]
+
+	glog.V(100).Infof("Creating new MachineSet copy of first existing worker MachineSet: %v",
+		baseMs.Name)
+
+	copiedMachineSet := &machinev1beta1.MachineSet{
+		ObjectMeta: *baseMs.ObjectMeta.DeepCopy(),
+		Spec:       *baseMs.Spec.DeepCopy(),
+	}
+
+	glog.V(100).Infof("Renaming copied MachineSet to: %s", copiedMachineSet.ObjectMeta.Name)
+
+	copiedMachineSet.ObjectMeta.Name = fmt.Sprintf("%v-%v", copiedMachineSet.Name,
+		strings.ReplaceAll(instanceType, ".", "-"))
+
+	glog.V(100).Infof("Updating copied MachineSet name in metadata, selector and template parameter ...")
```

Comment:
nit: what's the purpose of "<space>+ 3 dots" in the end of the message?

Reply:
No purpose, should not be there.  Will remove

Reply:
Done